### PR TITLE
CLOUDP-80014: refactor stateful set merging 1

### PR DESCRIPTION
**Pull request by chatton**
### All Submissions:

* [X] Have you signed our [CLA](https://www.mongodb.com/legal/contributor-agreement)?
* [X] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

This PR is the first part in a set of PRs that will add custom merging logic for merging StatefulSets.

### CODE REVIEW

File: pkg/kube/statefulset/merge_spec_test.go [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+			name: "Doesn't panic with nil input",
+			args: args{
+				original: nil,
+				override: nil,
+			},
+			want: []string{},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
```
Comment (bznein) [inline pkg/kube/statefulset/merge_spec_test.go:49]: Table Driven Tests 🥇

**Review by bznein — APPROVED** [review]
Great job 🥇 a few comments but nothing blocking!

File: pkg/kube/statefulset/merge_spec.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+	mergedLabelSelector := &metav1.LabelSelector{}
+	mergedLabelSelector.MatchLabels = mergeStringToStringMaps(originalLabelSelector.MatchLabels, overrideLabelSelector.MatchLabels)
+	mergedLabelSelector.MatchExpressions = mergeLabelSelectorRequirements(originalLabelSelector.MatchExpressions, overrideLabelSelector.MatchExpressions)
+
+	return mergedLabelSelector
+}
+
+// mergeStringToStringMaps returns a map containing all the keys of the original and override provided.
+// with any duplicate keys, values of the override will take precedence. A nil map is never returned.
+func mergeStringToStringMaps(original, override map[string]string) map[string]string {
```
[the remark this answers is not in the document]
Comment (chatton) [reply pkg/kube/statefulset/merge_spec.go:63]: I had the exact same though, in the next pr I'll be moving this into a `merge` package

File: pkg/kube/statefulset/merge_spec.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+		if lsr.Key == key {
+			return &lsr
+		}
+	}
+	return nil
+}
+
+// mergeDistinct accepts two slices of strings, and returns a string slice
+// containing the distinct elements present in both. The elements are returned sorted.
+func mergeDistinct(original, override []string) []string {
```
[the remark this answers is not in the document]
Comment (chatton) [reply pkg/kube/statefulset/merge_spec.go:128]: will be moving this one too!

File: pkg/kube/statefulset/merge_spec.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+	for _, lsr := range labelSelectorRequirements {
+		if lsr.Key == key {
+			return &lsr
+		}
+	}
+	return nil
+}
+
+// mergeDistinct accepts two slices of strings, and returns a string slice
+// containing the distinct elements present in both. The elements are returned sorted.
```
[the remark this answers is not in the document]
Comment (chatton) [reply pkg/kube/statefulset/merge_spec.go:127]: I think this might be an old comment, will make sure it's up to date.

**Pull request merged**